### ISSUE TITLE
Strange interaction with ArangoSearch view and filter

### ISSUE THREAD

**Issue by norpan** [arangodb/arangodb#8341]
## My Environment

* __ArangoDB Version__:       3.4.3
* __Storage Engine__:             RocksDB
* __Deployment Mode__:       Single Server
* __Deployment Strategy__:   manual start
* __Configuration__:               default settings
* __Infrastructure__:               
* __Operating System__:        Alpine 6.4.0
* __Total RAM in your machine__:        16GB
* __Disks in use__:        SSD
* __Used Package__:             Docker

## Component, Query & Data

__Affected feature__:
AQL Arangosearch

__AQL query (if applicable)__:
```
FOR movie in v_imdb
SEARCH movie.type == "Movie" and phrase(movie.description, "star", "text_en")
FOR actor in inbound movie imdb_edges
FILTER (FOR a in v_imdb SEARCH a._key == actor._key AND a.name == "Elizabeth Perkins" LIMIT 1 RETURN a) != []
FILTER actor.name == "Elizabeth Perkins"
return actor
```

__AQL explain (if applicable)__:
For first step:
```
Query String:
 FOR movie in v_imdb
 SEARCH movie.type == "Movie" and phrase(movie.description, "star", "text_en")
 FOR actor in inbound movie imdb_edges
 FILTER (FOR a in v_imdb SEARCH a._key == actor._key AND a.name == "Elizabeth Perkins" LIMIT 1 RETURN 
 a) != []
 FILTER actor.name == "Elizabeth Perkins"
 return actor

Execution plan:
 Id   NodeType            Est.   Comment
  1   SingletonNode          1   * ROOT
  2   EnumerateViewNode      1     - FOR movie IN v_imdb SEARCH ((movie.`type` == "Movie") && PHRASE(movie.`description`, "star", "text_en"))   /* view query */
  3   TraversalNode         13       - FOR actor  /* vertex */ IN 1..1  /* min..maxPathDepth */ INBOUND movie /* startnode */  imdb_edges
 11   CalculationNode       13         - LET #9 = (actor.`name` == "Elizabeth Perkins")   /* simple expression */
 12   FilterNode            13         - FILTER #9
  8   SubqueryNode          13         - LET #3 = ...   /* subquery */
  4   SingletonNode          1           * ROOT
  5   EnumerateViewNode      1             - FOR a IN v_imdb SEARCH ((a.`_key` == actor.`_key`) && (a.`name` == "Elizabeth Perkins"))   /* view query */
  6   LimitNode              1               - LIMIT 0, 1
  7   ReturnNode             1               - RETURN a
  9   CalculationNode       13         - LET #7 = (#3 != [ ])   /* simple expression */
 10   FilterNode            13         - FILTER #7
 13   ReturnNode            13         - RETURN actor

Indexes used:
 By   Type   Collection   Unique   Sparse   Selectivity   Fields      Ranges
  3   edge   imdb_edges   false    false            n/a   [ `_to` ]   base INBOUND

Traversals on graphs:
 Id   Depth   Vertex collections   Edge collections   Options                                   Filter conditions
  3   1..1                         imdb_edges         uniqueVertices: none, uniqueEdges: path   

Optimization rules applied:
 Id   RuleName
  1   move-calculations-up
  2   move-filters-up
  3   handle-arangosearch-views
```
For second step (with line 5 commented out)
```
Query String:
 FOR movie in v_imdb
 SEARCH movie.type == "Movie" and phrase(movie.description, "star", "text_en")
 FOR actor in inbound movie imdb_edges
 FILTER (FOR a in v_imdb SEARCH a._key == actor._key AND a.name == "Elizabeth Perkins" LIMIT 1 RETURN 
 a) != []
 //FILTER actor.name == "Elizabeth Perkins"
 return actor

Execution plan:
 Id   NodeType            Est.   Comment
  1   SingletonNode          1   * ROOT
  2   EnumerateViewNode      1     - FOR movie IN v_imdb SEARCH ((movie.`type` == "Movie") && PHRASE(movie.`description`, "star", "text_en"))   /* view query */
  3   TraversalNode         13       - FOR actor  /* vertex */ IN 1..1  /* min..maxPathDepth */ INBOUND movie /* startnode */  imdb_edges
  8   SubqueryNode          13         - LET #3 = ...   /* subquery */
  4   SingletonNode          1           * ROOT
  5   EnumerateViewNode      1             - FOR a IN v_imdb SEARCH ((a.`_key` == actor.`_key`) && (a.`name` == "Elizabeth Perkins"))   /* view query */
  6   LimitNode              1               - LIMIT 0, 1
  7   ReturnNode             1               - RETURN a
  9   CalculationNode       13         - LET #7 = (#3 != [ ])   /* simple expression */
 10   FilterNode            13         - FILTER #7
 11   ReturnNode            13         - RETURN actor

Indexes used:
 By   Type   Collection   Unique   Sparse   Selectivity   Fields      Ranges
  3   edge   imdb_edges   false    false            n/a   [ `_to` ]   base INBOUND

Traversals on graphs:
 Id   Depth   Vertex collections   Edge collections   Options                                   Filter conditions
  3   1..1                         imdb_edges         uniqueVertices: none, uniqueEdges: path   

Optimization rules applied:
 Id   RuleName
  1   handle-arangosearch-views
```

__Dataset__:
Dataset is the IMDB dataset described in the "Hands on tutorial" at https://www.arangodb.com/arangodb-training-center/search/arangosearch/

The view was created like this:
```
var v = db._createView("v_imdb", "arangosearch");
var link = { 
  includeAllFields: true,
  fields : { description : { analyzers : [ "text_en" ] } }
};
v.properties({ links: { imdb_vertices: link }})
```

## Steps to reproduce

1. Run the query
2. Comment out or remove line 5 of the query
3. Run the query again

__Problem__:
The first query returns "Elizabeth Perkins" (twice, since she is in two movies)
The second query returns an empty result.

__Expected result__:
The second query should return the same result as the first one.
It seems obviously wrong to me that removing a filter gives fewer results.

This was an attempt to filter using arangosearch matching on stemming etc, since that does not seem to be available using FILTER. It seems that the filter is moved up (which is fine of course), but also that this makes the filter using the search view being applied on fewer nodes(?) and so avoids some kind of limit.

Looking at the profiles, it looks like `EnumerateViewNode` finds 2 out of 2 in the first case, but 0 out of 9589 in the second case.

__Profiles__:
Profile with line 5:
```
Query String:
 FOR movie in v_imdb
 SEARCH movie.type == "Movie" and phrase(movie.description, "star", "text_en")
 FOR actor in inbound movie imdb_edges
 FILTER (FOR a in v_imdb SEARCH a._key == actor._key AND a.name == "Elizabeth Perkins" LIMIT 1 RETURN 
 a) != []
 FILTER actor.name == "Elizabeth Perkins"
 return actor

Execution plan:
 Id   NodeType            Calls   Items   Runtime [s]   Comment
  1   SingletonNode           1       1       0.00000   * ROOT
  2   EnumerateViewNode       1     488       0.00200     - FOR movie IN v_imdb SEARCH ((movie.`type` == "Movie") && PHRASE(movie.`description`, "star", "text_en"))   /* view query */
  3   TraversalNode          10    9589       0.06650       - FOR actor  /* vertex */ IN 1..1  /* min..maxPathDepth */ INBOUND movie /* startnode */  imdb_edges
 11   CalculationNode        10    9589       0.00256         - LET #9 = (actor.`name` == "Elizabeth Perkins")   /* simple expression */
 12   FilterNode              1       2       0.00133         - FILTER #9
  8   SubqueryNode            1       2       0.00011         - LET #3 = ...   /* subquery */
  4   SingletonNode           2       2       0.00000           * ROOT
  5   EnumerateViewNode       2       2       0.00009             - FOR a IN v_imdb SEARCH ((a.`_key` == actor.`_key`) && (a.`name` == "Elizabeth Perkins"))   /* view query */
  6   LimitNode               2       2       0.00000               - LIMIT 0, 1
  7   ReturnNode              2       2       0.00000               - RETURN a
  9   CalculationNode         1       2       0.00001         - LET #7 = (#3 != [ ])   /* simple expression */
 10   FilterNode              1       2       0.00000         - FILTER #7
 13   ReturnNode              1       2       0.00000         - RETURN actor

Indexes used:
 By   Type   Collection   Unique   Sparse   Selectivity   Fields      Ranges
  3   edge   imdb_edges   false    false            n/a   [ `_to` ]   base INBOUND

Traversals on graphs:
 Id   Depth   Vertex collections   Edge collections   Options                                   Filter conditions
  3   1..1                         imdb_edges         uniqueVertices: none, uniqueEdges: path   

Optimization rules applied:
 Id   RuleName
  1   move-calculations-up
  2   move-filters-up
  3   handle-arangosearch-views

Query Statistics:
 Writes Exec   Writes Ign   Scan Full   Scan Index   Filtered   Exec Time [s]
           0            0           0        10079       9587         0.07323

Query Profile:
 Query Stage           Duration [s]
 initializing               0.00000
 parsing                    0.00008
 optimizing ast             0.00001
 loading collections        0.00001
 instantiating plan         0.00004
 optimizing plan            0.00017
 executing                  0.07253
 finalizing                 0.00039
```
Profile without line 5:
```
Query String:
 FOR movie in v_imdb
 SEARCH movie.type == "Movie" and phrase(movie.description, "star", "text_en")
 FOR actor in inbound movie imdb_edges
 FILTER (FOR a in v_imdb SEARCH a._key == actor._key AND a.name == "Elizabeth Perkins" LIMIT 1 RETURN 
 a) != []
 //FILTER actor.name == "Elizabeth Perkins"
 return actor

Execution plan:
 Id   NodeType            Calls   Items   Runtime [s]   Comment
  1   SingletonNode           1       1       0.00000   * ROOT
  2   EnumerateViewNode       1     488       0.00191     - FOR movie IN v_imdb SEARCH ((movie.`type` == "Movie") && PHRASE(movie.`description`, "star", "text_en"))   /* view query */
  3   TraversalNode          10    9589       0.05965       - FOR actor  /* vertex */ IN 1..1  /* min..maxPathDepth */ INBOUND movie /* startnode */  imdb_edges
  8   SubqueryNode           10    9589       0.06222         - LET #3 = ...   /* subquery */
  4   SingletonNode        9589    9589       0.00577           * ROOT
  5   EnumerateViewNode    9589       0       0.02709             - FOR a IN v_imdb SEARCH ((a.`_key` == actor.`_key`) && (a.`name` == "Elizabeth Perkins"))   /* view query */
  6   LimitNode            9589       0       0.00932               - LIMIT 0, 1
  7   ReturnNode           9589       0       0.00819               - RETURN a
  9   CalculationNode        10    9589       0.00282         - LET #7 = (#3 != [ ])   /* simple expression */
 10   FilterNode              1       0       0.00153         - FILTER #7
 11   ReturnNode              1       0       0.00000         - RETURN actor

Indexes used:
 By   Type   Collection   Unique   Sparse   Selectivity   Fields      Ranges
  3   edge   imdb_edges   false    false            n/a   [ `_to` ]   base INBOUND

Traversals on graphs:
 Id   Depth   Vertex collections   Edge collections   Options                                   Filter conditions
  3   1..1                         imdb_edges         uniqueVertices: none, uniqueEdges: path   

Optimization rules applied:
 Id   RuleName
  1   handle-arangosearch-views

Query Statistics:
 Writes Exec   Writes Ign   Scan Full   Scan Index   Filtered   Exec Time [s]
           0            0           0        10077       9589         0.12889

Query Profile:
 Query Stage           Duration [s]
 initializing               0.00000
 parsing                    0.00008
 optimizing ast             0.00001
 loading collections        0.00001
 instantiating plan         0.00003
 optimizing plan            0.00020
 executing                  0.12817
 finalizing                 0.00040
```

**Comment by norpan** [comment]
I can confirm that this one is fixed in the latest 3.4 nightly! Thanks for the responses!

**Comment by Simran-B** [comment]
For everyone else: the issue was discussed on Slack and appears to be caused by an already fixed bug:
https://github.com/arangodb/arangodb/issues/8213

**Comment by maxkernbach** [comment]
Hi @norpan,

version 3.4.4, which includes a fix for this issue (the same fix that you already tested in the 3.4 nightly), has been released.
In case you are still facing this issue using v3.3.4+, please comment and we will reopen.
